snow/consensus/snowman/poll/prefix.go

```diff
@@ -0,0 +1,189 @@
+)
+
+// prefixGroup represents a bunch of IDs (stored in the members field),
+// with a bit prefix.
+// Each time the prefixGroup is split, it is divided into one or more prefixGroups
+// according to the next bit index in the index field.
+// Successively splitting prefixGroups yields a graph, with the first prefixGroup as the root.
+type prefixGroup struct {
+	// the bit index this prefixGroup would be split on by the next invocation of split().
+	index int
+	// the bits of the members of this prefixGroup from the first bit to the bit index.
+	prefix []uint8
+	// the IDs of the prefixGroup
+	members []ids.ID
+	// prefixGroups that correspond to zero and one being the first bit of their members, respectively.
+	zg, og *prefixGroup
+	// the prefixGroup that this prefixGroup was split from.
+	parent *prefixGroup
+	// was this prefixGroup split before. Used to prevent a prefixGroup from being split more than once,
+	// otherwise longestSharedPrefixes() would run indefinitely.
+	wasSplit bool
+}
+
+// longestSharedPrefixes creates a prefixGroup that is the root of a graph
+// of prefixGroup vertices.
+// When iterating the graph, each prefixGroup vertex represents a shared bit prefix
+// of IDs, and the members field contains all IDs from the given idList for which their bit prefix
+// matches the prefix field.
+func longestSharedPrefixes(idList []ids.ID) *prefixGroup {
+	originPG := &prefixGroup{members: idList}
+
+	pgs := make([]*prefixGroup, 0, len(idList))
+	pgs = append(pgs, originPG)
+
+	// Try to split each prefix group.
+	// Continue until all prefix groups cannot be split anymore.
+	for {
+		var couldSplit bool
+		for _, pg := range pgs {
+			if !pg.canSplit() {
+				continue
+			}
+
+			couldSplit = true
+
+			pg0, pg1 := pg.split()
+
+			if pg0 != nil && pg1 != nil {
+				pgs = append(pgs, pg0)
+				pgs = append(pgs, pg1)
+				continue
+			}
+
+			// Else, there is no bifurcation,
+			// so swallow up your descendant
+			descendant := determineDescendant(pg)
+
+			// Backup the parent
+			parent := pg.parent
+			// Become your descendant, but keep your original parent
+			*pg = *descendant
+			pg.parent = parent
+		}
+
+		if !couldSplit {
+			break
+		}
+	}
+
+	return originPG
+}
+
+func determineDescendant(pg *prefixGroup) *prefixGroup {
+	var descendant *prefixGroup
+
+	if pg.zg == nil && pg.og == nil {
+		// If both are nil, it's a programming error, so panic.
+		panic("programming error: both zero group and one group are nil")
+	}
+
+	if pg.zg != nil {
+		descendant = pg.zg
+	}
+
+	if pg.og != nil {
+		descendant = pg.og
+	}
+	return descendant
+}
+
+// bifurcationsWithCommonPrefix traverses the transitive descendants of this prefix group,
+// and applies f() on the block IDs of each prefix group.
+// Prefix groups with no descendants are skipped, as they do not represent any prefix.
+// Prefix group without a prefix (root prefix group) are also skipped as they do not correspond
+// to any instance of snowflake.
```

Comment:
Can we expand on this comment? Every bifurcation should correspond to a binary instance of snowflake no?

Reply:
Yes but here we're discussing actual blocks, not prefixes. A "root" prefix group doesn't have a prefix, because it corresponds to a real block. 

We count real blocks in other places in the code (`computeTransitiveVoteCountGraph`) so we shouldn't count these vertices twice, hence we skip them.

Reply:
Did you mean expand in the code as a comment or here?